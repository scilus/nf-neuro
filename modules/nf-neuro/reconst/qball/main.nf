process RECONST_QBALL {
    tag "$meta.id"
    label 'process_medium'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(dwi), path(bval), path(bvec), path(mask)

    output:
        tuple val(meta), path("*__gfa.nii.gz")           , emit: gfa, optional: true
        tuple val(meta), path("*__peaks.nii.gz")        , emit: peaks, optional: true
        tuple val(meta), path("*__peak_indices.nii.gz")        , emit: peak_indices, optional: true
        tuple val(meta), path("*__sh.nii.gz")       , emit: sh, optional: true
        tuple val(meta), path("*__nufo.nii.gz")             , emit: nufo, optional: true
        tuple val(meta), path("*__a_power.nii.gz")         , emit: a_power, optional: true
        path "versions.yml"                                , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def set_mask = task.ext.mask ? "--mask $mask" : ""
    def b0_threshold = task.ext.b0_threshold ? "--b0_threshold " + task.ext.b0_threshold : ""

    if ( task.ext.gfa ) gfa = "--gfa_fODF ${prefix}__gfa.nii.gz" else gfa = ""
    if ( task.ext.peaks ) peaks = "--peaks ${prefix}__peaks.nii.gz" else peaks = ""
    if ( task.ext.peak_indices ) peak_indices = "--peak_indices ${prefix}__peak_indices.nii.gz" else peak_indices = ""
    if ( task.ext.sh) sh = "--sh ${prefix}__sh.nii.gz" else sh = ""
    if ( task.ext.nufo) nufo = "--nufo ${prefix}__nufo.nii.gz" else nufo = ""
    if ( task.ext.a_power) a_power = "--a_power ${prefix}__a_power.nii.gz" else a_power = ""
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_qball_metrics.py $dwi $bval $bvec $set_mask --not_all \
    $gfa $peaks $peak_indices $sh $nufo $a_power $b0_threshold
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_qball_metrics.py -h

    touch ${prefix}__gfa.nii.gz
    touch ${prefix}__peaks.nii.gz
    touch ${prefix}__peak_indices.nii.gz
    touch ${prefix}__sh.nii.gz
    touch ${prefix}__nufo.nii.gz
    touch ${prefix}__a_power.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
