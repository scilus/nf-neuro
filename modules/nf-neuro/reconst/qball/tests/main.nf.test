// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
// nf-core modules test reconst/qball
nextflow_process {

    name "Test Process RECONST_QBALL"
    script "../main.nf"
    process "RECONST_QBALL"

    tag "modules"
    tag "modules_nfcore"
    tag "reconst"
    tag "reconst/qball"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [
                    "DWIss2000-dir60.zip",
                    "DWIms1000-2000-dir10-10.zip",
                    "segmentation.zip",
                    "responses.zip"
                ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }
    test("reconst - qball") {
        config "./nextflow_msmt.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        ssst: it.simpleName == "DWIss2000-dir60"
                        segmentation: it.simpleName == "segmentation"
                    }
                ch_dwi = ch_split_test_data.ssst.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]
                }
                ch_mask = ch_split_test_data.segmentation.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/brainmask/brainmask.nii.gz")
                    ]
                }
                input[0] = ch_dwi
                    .join(ch_mask)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("reconst - qball_gfa") {
        config "./nextflow_light"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        ssst: it.simpleName == "DWIss2000-dir60"
                        segmentation: it.simpleName == "segmentation"
                    }
                ch_dwi = ch_split_test_data.ssst.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]
                }
                ch_mask = ch_split_test_data.segmentation.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        []
                    ]
                }
                input[0] = ch_dwi
                    .join(ch_mask)
                """
            }
            then {
                assertAll(
                    { assert process.success },
                    { assert snapshot(process.out).match() }
                )
            }

        }
    }

test("reconst - stub-run") {
        options "-stub-run"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        ssst: it.simpleName == "DWIss2000-dir60"
                        segmentation: it.simpleName == "segmentation"
                    }
                ch_dwi = ch_split_test_data.ssst.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]
                }
                ch_mask = ch_split_test_data.segmentation.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/brainmask/brainmask.nii.gz")
                    ]
                }
                input[0] = ch_dwi
                    .join(ch_mask)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }
}
