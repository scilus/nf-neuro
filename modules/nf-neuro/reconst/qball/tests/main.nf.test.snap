{
    "reconst - stub-run": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__gfa.nii.gz"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__peaks.nii.gz"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__peak_indices.nii.gz"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__sh.nii.gz"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__nufo.nii.gz"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__a_power.nii.gz"
                    ]
                ],
                "6": [
                    "versions.yml:md5,ed8e825c5eccf94551ba3af620020a94"
                ],
                "a_power": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__a_power.nii.gz"
                    ]
                ],
                "gfa": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__gfa.nii.gz"
                    ]
                ],
                "nufo": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__nufo.nii.gz"
                    ]
                ],
                "peak_indices": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__peak_indices.nii.gz"
                    ]
                ],
                "peaks": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__peaks.nii.gz"
                    ]
                ],
                "sh": [
                    [
                        {
                            "id": "test"
                        },
                        "/workspaces/nf-neuro/tests/.runs/tests/738f74515d0739ddd47874c7107fec42/work/1c/69722101bf73f201b5518371bb0ecc/test__sh.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ed8e825c5eccf94551ba3af620020a94"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-27T14:46:54.341368697"
    }
}